lib/mix/tasks/benchmark.ex

```diff
@@ -0,0 +1,199 @@
+    |> compile_benchmark_files(opts)
+  end
+
+  defp find_benchmark_files(:stop, _opts), do: :stop
+
+  defp find_benchmark_files(files, _opts) do
+    :ok = Logger.debug("Locating benchmarks")
+
+    files =
+      case files do
+        [] ->
+          [default_benchmark_path()]
+
+        o ->
+          o
+      end
+
+    files =
+      Enum.flat_map(files, fn file ->
+        if File.dir?(file) do
+          [file, "*.exs"]
+          |> Path.join()
+          |> Path.wildcard()
+        else
+          [file]
+        end
+      end)
+
+    case files do
+      [] ->
+        :ok = Logger.debug("No benchmark files found")
+        :stop
+
+      files ->
+        files
+    end
+  end
+
+  defp compile_benchmark_files(:stop, _opts), do: :stop
+
+  defp compile_benchmark_files(files, _opts) do
+    :ok = Logger.debug("Compiling benchmarks")
+
+    modules =
+      files
+      |> Enum.flat_map(fn f ->
+        f
+        |> Helper.compile_file()
+        |> Enum.map(fn {mod, _bin} -> mod end)
+      end)
+
+    case modules do
+      [] ->
+        :ok = Logger.debug("No benchmarks found")
+        :stop
+
+      modules ->
+        modules
+    end
+  end
+
+  defp run_benchmarks(:stop, _opts), do: :stop
+
+  defp run_benchmarks(modules, _opts) do
+    :ok = Logger.debug("Running benchmarks")
+
+    modules
+    |> get_benchmarks()
+    |> run_all_benchmarks()
+  end
+
+  defp run_all_benchmarks(benchmarks, print_system? \\ true)
+  defp run_all_benchmarks([], _print?), do: :ok
+
+  defp run_all_benchmarks([{mod, run_funs} | rest], print_system?) do
+    opts =
+      case print_system? do
+        true -> []
+        false -> [print: [system: false]]
+      end
+
+    case apply_if_exists(mod, :setup, [], {:ok, nil}) do
+      {:ok, state} ->
+        state =
+          Enum.reduce(run_funs, state, fn funname, state ->
+            apply(mod, funname, [state, opts])
+          end)
+
+        try_apply_if_exists(mod, :teardown, [state])
+
+      other ->
+        :ok =
+          Logger.error(
+            "Global setup of \"#{inspect(mod)}\" " <>
+              "returned #{inspect(other)}"
+          )
+    end
+
+    run_all_benchmarks(rest, false)
+  end
+
+  defp try_apply_if_exists(mod, name, args) do
+    case apply_if_exists(mod, name, args) do
+      _ -> :ok
+    end
+  end
+
+  defp apply_if_exists(mod, name, args, default) do
+    case apply_if_exists(mod, name, args) do
+      {:ok, result} -> result
+      {:error, :not_exists} -> default
+    end
+  end
+
+  defp apply_if_exists(mod, fun, args) do
+    arity = length(args)
+
+    case mod.module_info(:exports)[fun] do
+      ^arity -> {:ok, apply(mod, fun, args)}
+      nil -> {:error, :not_exists}
+    end
+  end
+
+  defp get_benchmarks(modules) do
+    case_attr = Benchee.Project.Benchmark.benchee_scenario_attr()
+
+    modules
+    |> Stream.filter(fn mod ->
+      Keyword.has_key?(mod.module_info(:attributes), case_attr)
+    end)
+    |> Stream.map(fn mod ->
```

Comment:
Do we really need to go down to `Stream` here for just 2 relatively simple invocations? We're calling `to_list` right at the end so I'm not sure it's necessary

Reply:
Actually, I'll try to rewrite it as a list comprehension.

Reply:
Personally I usually prefer `Enum` calls over list comprehensions and the transfer would be quite straight forward :grin:

Reply:
Now I finally understand your remark.
I can imagine rather big benchmarks with a lot of helper modules, may be mocks and what else. So I think Stream might be more idiomatic here. Although I hardly imagine benchmarks with more than, say, 1000 modules, so Stream might actually be unnecessary here.

Will change to Enum.